File: Source/Items/Items.c
/****************************/
/*   		ITEMS.C		    */
/* (c)2003 Pangea Software  */
/* By Brian Greenstone      */
/****************************/


/****************************/
/*    EXTERNALS             */
/****************************/


#include "game.h"


/****************************/
/*    PROTOTYPES            */
/****************************/

static void CreateCloudLayer(void);
static void DrawCloudLayer(ObjNode *theNode);
static void MoveCloudLayer(ObjNode *theNode);
static void MoveGasMound(ObjNode *theNode);



/****************************/
/*    CONSTANTS             */
/****************************/


#define	LEAF_DEFAULT_WOBBLE_MAG		4.0f
#define	LEAF_DEFAULT_WOBBLE_SPEED	1.0f

/*********************/
/*    VARIABLES      */
/*********************/


static OGLVector2D	gCloudScroll;


/********************* INIT ITEMS MANAGER *************************/

void InitItemsManager(void)
{
	InitForestDoors();
	InitZaps();
	InitWormholes();
	InitDustDevilMemory();

	CreateCyclorama();
	CreateCloudLayer();

			/* UPDATE SONG */

//	if (gSongMovie)
//		MoviesTask(gSongMovie, 0);

}


/************************* CREATE CYCLORAMA *********************************/

void CreateCyclorama(void)
{
	NewObjectDefinitionType def =
	{
		.group		= MODEL_GROUP_LEVELSPECIFIC,
		.type		= 0,								// cyc is always 1st model in level bg3d files
		.coord		= {0,0,0},
		.flags		= STATUS_BIT_DONTCULL|STATUS_BIT_NOLIGHTING|STATUS_BIT_NOFOG, //|STATUS_BIT_NOZWRITES;
		.slot		= TERRAIN_SLOT+1,					// draw after terrain for better performance since terrain blocks much of the pixels
		.moveCall	= nil,
		.drawCall	= DrawCyclorama,
		.rot		= 0,
		.scale		= gGameViewInfoPtr->yon * .01,
	};

	MakeNewDisplayGroupObject(&def);
}


/********************** DRAW CYCLORAMA *************************/

void DrawCyclorama(ObjNode *theNode)
{
OGLPoint3D cameraCoord = gGameViewInfoPtr->cameraPlacement[gCurrentSplitScreenPane].cameraLocation;

	glDisable(GL_ALPHA_TEST);	//--------


		/* UPDATE CYCLORAMA COORD INFO */

	theNode->Coord.x = cameraCoord.x;
	theNode->Coord.y = cameraCoord.y;
	theNode->Coord.z = cameraCoord.z;
	UpdateObjectTransforms(theNode);


			/* DRAW THE OBJECT */

	MO_DrawObject(theNode->BaseGroup);

	glEnable(GL_ALPHA_TEST);	//--------

}




/************************* CREATE CLOUD LAYER *********************************/

static void CreateCloudLayer(void)
{
	if (gGamePrefs.lowRenderQuality)			// don't do cloulds in low-quality mode
		return;

	NewObjectDefinitionType def =
	{
		.group		= MODEL_GROUP_LEVELSPECIFIC,
		.type		= LEVEL1_ObjType_CloudGrid,
		.coord		= {0,0,0},
		.flags		= STATUS_BIT_DONTCULL|STATUS_BIT_NOLIGHTING|STATUS_BIT_NOFOG| STATUS_BIT_DOUBLESIDED | STATUS_BIT_NOZWRITES,
		.slot		= TERRAIN_SLOT+2,					 // draw after sky dome
		.rot		= 0,
		.scale		= gGameViewInfoPtr->yon * .85f,
		.moveCall	= MoveCloudLayer,
		.drawCall	= DrawCloudLayer,
	};

	ObjNode* newObj = MakeNewDisplayGroupObject(&def);

	newObj->TextureTransformU =
	newObj->TextureTransformV = 0;
}


/******************* MOVE CLOUD LAYER *********************/

static void MoveCloudLayer(ObjNode *theNode)
{
#pragma unused (theNode)

	gCloudScroll.x += gFramesPerSecondFrac * .02f;
	gCloudScroll.y += gFramesPerSecondFrac * .03f;
}


/********************** DRAW CLOUD LAYER *************************/

static void DrawCloudLayer(ObjNode *theNode)
{
OGLPoint3D cameraCoord = gGameViewInfoPtr->cameraPlacement[gCurrentSplitScreenPane].cameraLocation;

	glDisable(GL_ALPHA_TEST);	//--------

		/* UPDATE CYCLORAMA COORD INFO */

	theNode->Coord.x = cameraCoord.x;
	theNode->Coord.y = 4000;
	theNode->Coord.z = cameraCoord.z;
	UpdateObjectTransforms(theNode);

		/* UPDATE SCROLLING */

	theNode->TextureTransformU = cameraCoord.x * .0001 + gCloudScroll.x;
	theNode->TextureTransformV = cameraCoord.z * -.0001 + gCloudScroll.y;

	glMatrixMode(GL_TEXTURE);					// set texture matrix
	glTranslatef(theNode->TextureTransformU, theNode->TextureTransformV, 0);
	glMatrixMode(GL_MODELVIEW);


			/* DRAW THE OBJECT */

	MO_DrawObject(theNode->BaseGroup);



				/* RESET UV MATRIX */

	glMatrixMode(GL_TEXTURE);
	glLoadIdentity();
	glMatrixMode(GL_MODELVIEW);

	glEnable(GL_ALPHA_TEST);	//--------

}


#pragma mark -




/************************* ADD ROCK *********************************/

Boolean AddRock(TerrainItemEntryType *itemPtr, float  x, float z)
{
short	base;
long	rot = itemPtr->parm[1];

	switch(gLevelNum)
	{
		case	LEVEL_NUM_ADVENTURE1:
		case	LEVEL_NUM_FLAG2:
		case	LEVEL_NUM_BATTLE1:
				base = 	LEVEL1_ObjType_Rock1;
				break;

		case	LEVEL_NUM_ADVENTURE2:
		case	LEVEL_NUM_RACE2:
		case	LEVEL_NUM_BATTLE2:
				base = 	LEVEL2_ObjType_Rock_Small1;
				break;

		default:
				DoFatalAlert("AddRock: brian needs to assign rocks to this level");
				return(false);
	}

	NewObjectDefinitionType def =
	{
		.group 		= MODEL_GROUP_LEVELSPECIFIC,
		.type 		= base + itemPtr->parm[0],
		.scale 		= 2.0 + RandomFloat2() * .3f,
		.coord.x 	= x,
		.coord.z 	= z,
		.flags 		= gAutoFadeStatusBits,
		.slot 		= 491,
		.moveCall 	= MoveStaticObject,
		.rot		= (rot == 0) ? (RandomFloat()*PI2) : ((float)(rot-1) * (PI2/8.0f)),
	};
	def.coord.y 	= GetMinTerrainY(x,z, def.group, def.type, def.scale) - gObjectGroupBBoxList[def.group][def.type].min.y;

	ObjNode* newObj = MakeNewDisplayGroupObject(&def);

	newObj->TerrainItemPtr = itemPtr;								// keep ptr to item list



			/* SET COLLISION STUFF */

	newObj->CType 			= CTYPE_SOLIDTOENEMY | CTYPE_PLAYERTEST | CTYPE_WEAPONTEST;
	newObj->CBits			= CBITS_ALLSOLID;
	CreateCollisionBoxFromBoundingBox_Rotated(newObj, .7, .8);


	return(true);													// item was added
}

/************************* ADD RIVER ROCK *********************************/

Boolean AddRiverRock(TerrainItemEntryType *itemPtr, float  x, float z)
{
	NewObjectDefinitionType def =
	{
		.group 		= MODEL_GROUP_LEVELSPECIFIC,
		.type 		= LEVEL1_ObjType_RiverRock1 + itemPtr->parm[0],
		.scale 		= 2.0 + RandomFloat2() * .3f,
		.coord.x 	= x,
		.coord.z 	= z,
		.flags 		= gAutoFadeStatusBits,
		.slot 		= 491,
		.moveCall 	= MoveStaticObject,
		.rot 		= RandomFloat()*PI2,
	};

	def.coord.y 	= GetMinTerrainY(x,z, def.group, def.type, def.scale) - gObjectGroupBBoxList[def.group][def.type].min.y;

	ObjNode* newObj = MakeNewDisplayGroupObject(&def);

	newObj->TerrainItemPtr = itemPtr;								// keep ptr to item list



			/* SET COLLISION STUFF */

	newObj->CType 			= CTYPE_SOLIDTOENEMY | CTYPE_PLAYERTEST | CTYPE_WEAPONTEST;
	newObj->CBits			= CBITS_ALLSOLID;
	CreateCollisionBoxFromBoundingBox_Rotated(newObj, 1, 1);


	return(true);													// item was added
}


#pragma mark -


/************************* ADD GAS MOUND *********************************/

Boolean AddGasMound(TerrainItemEntryType *itemPtr, float  x, float z)
{
	NewObjectDefinitionType def =
	{
		.group 		= MODEL_GROUP_LEVELSPECIFIC,
		.type 		= LEVEL1_ObjType_GasMound1 + itemPtr->parm[0],
		.scale 		= 3.0 + RandomFloat2() * .3f,
		.coord.x 	= x,
		.coord.z 	= z,
		.flags 		= gAutoFadeStatusBits,
		.slot 		= 197,
		.moveCall 	= MoveGasMound,
		.rot 		= RandomFloat()*PI2,
	};

	def.coord.y 	= GetMinTerrainY(x,z, def.group, def.type, def.scale) - gObjectGroupBBoxList[def.group][def.type].min.y;

	ObjNode* newObj = MakeNewDisplayGroupObject(&def);

	newObj->TerrainItemPtr = itemPtr;								// keep ptr to item list

	newObj->Kind = itemPtr->parm[0];

			/* SET COLLISION STUFF */

	newObj->CType 			= CTYPE_SOLIDTOENEMY | CTYPE_PLAYERTEST | CTYPE_WEAPONTEST;
	newObj->CBits			= CBITS_ALLSOLID;
	CreateCollisionBoxFromBoundingBox_Rotated(newObj, 1, 1);


	return(true);													// item was added
}



/****************** MOVE GAS MOUND ****************************/

static void MoveGasMound(ObjNode *theNode)
{
int		i;
float	fps = gFramesPerSecondFrac;
int		particleGroup,magicNum;
NewParticleGroupDefType	groupDef;
NewParticleDefType	newParticleDef;
OGLVector3D			d;
OGLPoint3D			p;

	if (TrackTerrainItem(theNode))							// just check to see if it's gone
	{
		DeleteObject(theNode);
		return;
	}



	/* SEE IF SMOKE HITS PLAYER */

	for (i = 0; i < gNumPlayers; i++)
	{
		if (!gPlayerIsDead[i])								// ingore dead players
		{
			float   dist = CalcQuickDistance(theNode->Coord.x, theNode->Coord.z, gPlayerInfo[i].coord.x, gPlayerInfo[i].coord.z);

			if (dist < 150.0f)
			{
				dist = gPlayerInfo[i].coord.y - theNode->Coord.y;
				if (dist < 700.0f)
				{
					if (gPlayerInfo[i].objNode->Skeleton->AnimNum != PLAYER_ANIM_DISORIENTED)		// play effect on 1st hit
						PlayEffect3D(EFFECT_BODYHIT, &gPlayerInfo[i].coord);

					PlayerLoseHealth(i, fps * .1f, PLAYER_DEATH_TYPE_DEATHDIVE, nil, true);
				}
			}
		}
	}


		/**************/
		/* MAKE SMOKE */
		/**************/

	theNode->ParticleTimer -= fps;											// see if add smoke
	if (theNode->ParticleTimer <= 0.0f)
	{
		theNode->ParticleTimer += .05;										// reset timer

		particleGroup 	= theNode->ParticleGroup;
		magicNum 		= theNode->ParticleMagicNum;

		if ((particleGroup == -1) || (!VerifyParticleGroupMagicNum(particleGroup, magicNum)))
		{
			theNode->ParticleMagicNum = magicNum = MyRandomLong();			// generate a random magic num

			groupDef.magicNum				= magicNum;
			groupDef.type					= PARTICLE_TYPE_FALLINGSPARKS;
			groupDef.flags					= PARTICLE_FLAGS_DONTCHECKGROUND;
			groupDef.gravity				= 100;
			groupDef.magnetism				= 0;
			groupDef.baseScale				= 15.0f;
			groupDef.decayRate				=  -2.5f;
			groupDef.fadeRate				= .25;
			groupDef.particleTextureNum		= PARTICLE_SObjType_GasCloud;
			groupDef.srcBlend				= GL_SRC_ALPHA;
			groupDef.dstBlend				= GL_ONE_MINUS_SRC_ALPHA;
			theNode->ParticleGroup = particleGroup = NewParticleGroup(&groupDef);
		}

		if (particleGroup != -1)
		{
			float	x,y,z;
			const OGLPoint3D ventOff[3] =
			{
				{2.163, 125.091, -1.907},
				{19.385, 69.243, 2.243},
				{5.189, 22.168, -5.504}
			};

			OGLPoint3D_Transform(&ventOff[theNode->Kind], &theNode->BaseTransformMatrix, &p);

			x = p.x;
			y = p.y;
			z = p.z;

			for (i = 0; i < 2; i++)
			{
				p.x = x + RandomFloat2() * 15.0f;
				p.y = y + RandomFloat() * 10.0f;
				p.z = z + RandomFloat2() * 15.0f;

				d.x = RandomFloat2() * 30.0f;
				d.y = 300.0f + RandomFloat() * 150.0f;
				d.z = RandomFloat2() * 30.0f;

				newParticleDef.groupNum		= particleGroup;
				newParticleDef.where		= &p;
				newParticleDef.delta		= &d;
				newParticleDef.scale		= RandomFloat() + 1.0f;
				newParticleDef.rotZ			= RandomFloat() * PI2;
				newParticleDef.rotDZ		= RandomFloat2() * 2.0f;
				newParticleDef.alpha		= .6;
				if (AddParticleToGroup(&newParticleDef))
				{
					theNode->ParticleGroup = -1;
					break;
				}
			}
		}
	}


}



#pragma mark -

/******************** TRIGGER CALLBACK:  MISC SMACKABLE OBJECT ************************/
//
// Returns TRUE if want to handle hit as a solid
//

Boolean DoTrig_MiscSmackableObject(ObjNode *trigger, ObjNode *theNode)
{
	if (PlayerSmackedIntoObject(theNode, trigger, PLAYER_DEATH_TYPE_EXPLODE))
		return(false);

	return(true);
}



#pragma mark -

/************************* ADD ASTEROID *********************************/

Boolean AddAsteroid(TerrainItemEntryType *itemPtr, float  x, float z)
{
	NewObjectDefinitionType def =
	{
		.group 		= MODEL_GROUP_LEVELSPECIFIC,
		.type 		= LEVEL3_ObjType_Asteroid_Cracked + itemPtr->parm[0],
		.scale 		= 4.0 + RandomFloat2() * 1.0f,
		.coord.x 	= x,
		.coord.z 	= z,
		.flags 		= gAutoFadeStatusBits,
		.slot 		= 197,
		.moveCall 	= nil,
		.rot 		= RandomFloat()*PI2,
	};

	def.coord.y 	= GetMinTerrainY(x,z, def.group, def.type, def.scale) - gObjectGroupBBoxList[def.group][def.type].min.y;

	ObjNode* newObj = MakeNewDisplayGroupObject(&def);

	newObj->TerrainItemPtr = itemPtr;								// keep ptr to item list

	newObj->Kind = itemPtr->parm[0];

			/* SET COLLISION STUFF */

	newObj->CType 			= CTYPE_SOLIDTOENEMY | CTYPE_PLAYERTEST | CTYPE_WEAPONTEST;
	newObj->CBits			= CBITS_ALLSOLID;
	CreateCollisionBoxFromBoundingBox_Rotated(newObj, 1, 1);


	return(true);													// item was added
}

File: Source/Items/Items.swift
// Items.swift - Port of Items.c to Swift

private let leafDefaultWobbleMag: Float = 4.0
private let leafDefaultWobbleSpeed: Float = 1.0

private var gCloudScroll = OGLVector2D()

// MARK: - Init items manager

func InitItemsManager() {
    InitForestDoors()
    InitZaps()
    InitWormholes()
    InitDustDevilMemory()

    CreateCyclorama()
    createCloudLayer()

    // UPDATE SONG

    // if (gSongMovie)
    //     MoviesTask(gSongMovie, 0);
}

// MARK: - Create cyclorama

func CreateCyclorama() {
    var def = NewObjectDefinitionType()
    def.group = UInt8(MODEL_GROUP_LEVELSPECIFIC)
    def.type = 0 // cyc is always 1st model in level bg3d files
    def.coord = OGLPoint3D(x: 0, y: 0, z: 0)
    def.flags = UInt32(STATUS_BIT_DONTCULL | STATUS_BIT_NOLIGHTING | STATUS_BIT_NOFOG) // |STATUS_BIT_NOZWRITES;
    def.slot = Int16(TERRAIN_SLOT) + 1 // draw after terrain for better performance since terrain blocks much of the pixels
    def.moveCall = nil
    def.drawCall = DrawCyclorama
    def.rot = 0
    def.scale = gGameViewInfoPtr!.pointee.yon * 0.01

    _ = MakeNewDisplayGroupObject(&def)
}

// MARK: - Draw cyclorama

func DrawCyclorama(_ theNodeOpt: UnsafeMutablePointer<ObjNode>?) {
    let theNode = theNodeOpt!
    let cameraCoord = cameraPlacementsBase()[Int(gCurrentSplitScreenPane)].cameraLocation

    gRenderBackend.setAlphaTestEnabled(false) // --------

    // UPDATE CYCLORAMA COORD INFO

    theNode.pointee.Coord.x = cameraCoord.x
    theNode.pointee.Coord.y = cameraCoord.y
    theNode.pointee.Coord.z = cameraCoord.z
    theNode.updateTransforms()

    // DRAW THE OBJECT

    MO_DrawObject(theNode.pointee.BaseGroup)

    gRenderBackend.setAlphaTestEnabled(true) // --------
}

// cameraPlacement is a fixed-size array (imports as a tuple); rebind to a pointer so it can be dynamically indexed.
@inline(__always) private func cameraPlacementsBase() -> UnsafeMutablePointer<OGLCameraPlacement> {
    UnsafeMutableRawPointer(gGameViewInfoPtr!.pointer(to: \.cameraPlacement)!).assumingMemoryBound(to: OGLCameraPlacement.self)
}

// MARK: - Create cloud layer

private func createCloudLayer() {
    if gGamePrefs.isLowRenderQuality { // don't do clouds in low-quality mode
        return
    }

    var def = NewObjectDefinitionType()
    def.group = UInt8(MODEL_GROUP_LEVELSPECIFIC)
    def.type = UInt8(LEVEL1_ObjType_CloudGrid)
    def.coord = OGLPoint3D(x: 0, y: 0, z: 0)
    def.flags = UInt32(STATUS_BIT_DONTCULL | STATUS_BIT_NOLIGHTING | STATUS_BIT_NOFOG | STATUS_BIT_DOUBLESIDED | STATUS_BIT_NOZWRITES)
    def.slot = Int16(TERRAIN_SLOT) + 2 // draw after sky dome
    def.rot = 0
    def.scale = gGameViewInfoPtr!.pointee.yon * 0.85
    def.moveCall = cMoveCloudLayer
    def.drawCall = cDrawCloudLayer

    let newObj = MakeNewDisplayGroupObject(&def)!

    newObj.pointee.TextureTransformU = 0
    newObj.pointee.TextureTransformV = 0
}

// MARK: - Move cloud layer

private let cMoveCloudLayer: @convention(c) (UnsafeMutablePointer<ObjNode>?) -> Void = { _ in
    gCloudScroll.x += gFramesPerSecondFrac * 0.02
    gCloudScroll.y += gFramesPerSecondFrac * 0.03
}

// MARK: - Draw cloud layer

private let cDrawCloudLayer: @convention(c) (UnsafeMutablePointer<ObjNode>?) -> Void = { theNodeOpt in
    let theNode = theNodeOpt!
    let cameraCoord = cameraPlacementsBase()[Int(gCurrentSplitScreenPane)].cameraLocation

    gRenderBackend.setAlphaTestEnabled(false) // --------

    // UPDATE CYCLORAMA COORD INFO

    theNode.pointee.Coord.x = cameraCoord.x
    theNode.pointee.Coord.y = 4000
    theNode.pointee.Coord.z = cameraCoord.z
    theNode.updateTransforms()

    // UPDATE SCROLLING

    theNode.pointee.TextureTransformU = cameraCoord.x * 0.0001 + gCloudScroll.x
    theNode.pointee.TextureTransformV = cameraCoord.z * -0.0001 + gCloudScroll.y

    gRenderBackend.matrixMode(.texture) // set texture matrix
    gRenderBackend.translate(theNode.pointee.TextureTransformU, theNode.pointee.TextureTransformV, 0)
    gRenderBackend.matrixMode(.modelview)

    // DRAW THE OBJECT

    MO_DrawObject(theNode.pointee.BaseGroup)

    // RESET UV MATRIX

    gRenderBackend.matrixMode(.texture)
    gRenderBackend.loadIdentity()
    gRenderBackend.matrixMode(.modelview)

    gRenderBackend.setAlphaTestEnabled(true) // --------
}

// MARK: -

// MARK: - Add rock

extension UnsafeMutablePointer where Pointee == TerrainItemEntryType {
    @discardableResult
    func addRock(x: Float, z: Float) -> UInt8 {
        let base: Int32
        let rot = Int(pointee.parm.1)

        switch gLevelNum {
        case Int16(LevelNum.adventure1.rawValue), Int16(LevelNum.flag2.rawValue), Int16(LevelNum.battle1.rawValue):
            base = Int32(LEVEL1_ObjType_Rock1)

        case Int16(LevelNum.adventure2.rawValue), Int16(LevelNum.race2.rawValue), Int16(LevelNum.battle2.rawValue):
            base = Int32(LEVEL2_ObjType_Rock_Small1)

        default:
            SwFatal("AddRock: brian needs to assign rocks to this level")
            return 0
        }

        var def = NewObjectDefinitionType()
        def.group = UInt8(MODEL_GROUP_LEVELSPECIFIC)
        def.type = UInt8(base + Int32(pointee.parm.0))
        def.scale = 2.0 + RandomFloat2() * 0.3
        def.coord.x = x
        def.coord.z = z
        def.flags = gAutoFadeStatusBits
        def.slot = 491
        def.moveCall = MoveStaticObject
        def.rot = (rot == 0) ? (RandomFloat() * SwPI2) : (Float(rot - 1) * (SwPI2 / 8.0))

        def.coord.y = GetMinTerrainY(x, z, Int16(def.group), Int16(def.type), def.scale) - GetObjectGroupBBox(Int32(def.group), Int32(def.type)).min.y

        let newObj = MakeNewDisplayGroupObject(&def)!

        newObj.pointee.TerrainItemPtr = self // keep ptr to item list

        // SET COLLISION STUFF

        newObj.pointee.CType = UInt32(CTYPE_SOLIDTOENEMY | CTYPE_PLAYERTEST | CTYPE_WEAPONTEST)
        newObj.pointee.CBits = UInt32(CBITS_ALLSOLID)
        CreateCollisionBoxFromBoundingBox_Rotated(newObj, 0.7, 0.8)

        return 1 // item was added
    }
}

// MARK: - Add river rock

extension UnsafeMutablePointer where Pointee == TerrainItemEntryType {
    @discardableResult
    func addRiverRock(x: Float, z: Float) -> UInt8 {
        var def = NewObjectDefinitionType()
        def.group = UInt8(MODEL_GROUP_LEVELSPECIFIC)
        def.type = UInt8(Int32(LEVEL1_ObjType_RiverRock1) + Int32(pointee.parm.0))
        def.scale = 2.0 + RandomFloat2() * 0.3
        def.coord.x = x
        def.coord.z = z
        def.flags = gAutoFadeStatusBits
        def.slot = 491
        def.moveCall = MoveStaticObject
        def.rot = RandomFloat() * SwPI2

        def.coord.y = GetMinTerrainY(x, z, Int16(def.group), Int16(def.type), def.scale) - GetObjectGroupBBox(Int32(def.group), Int32(def.type)).min.y

        let newObj = MakeNewDisplayGroupObject(&def)!

        newObj.pointee.TerrainItemPtr = self // keep ptr to item list

        // SET COLLISION STUFF

        newObj.pointee.CType = UInt32(CTYPE_SOLIDTOENEMY | CTYPE_PLAYERTEST | CTYPE_WEAPONTEST)
        newObj.pointee.CBits = UInt32(CBITS_ALLSOLID)
        CreateCollisionBoxFromBoundingBox_Rotated(newObj, 1, 1)

        return 1 // item was added
    }
}

// MARK: -

// MARK: - Add gas mound

extension UnsafeMutablePointer where Pointee == TerrainItemEntryType {
    @discardableResult
    func addGasMound(x: Float, z: Float) -> UInt8 {
        var def = NewObjectDefinitionType()
        def.group = UInt8(MODEL_GROUP_LEVELSPECIFIC)
        def.type = UInt8(Int32(LEVEL1_ObjType_GasMound1) + Int32(pointee.parm.0))
        def.scale = 3.0 + RandomFloat2() * 0.3
        def.coord.x = x
        def.coord.z = z
        def.flags = gAutoFadeStatusBits
        def.slot = 197
        def.moveCall = cMoveGasMound
        def.rot = RandomFloat() * SwPI2

        def.coord.y = GetMinTerrainY(x, z, Int16(def.group), Int16(def.type), def.scale) - GetObjectGroupBBox(Int32(def.group), Int32(def.type)).min.y

        let newObj = MakeNewDisplayGroupObject(&def)!

        newObj.pointee.TerrainItemPtr = self // keep ptr to item list

        newObj.pointee.Kind = Int32(pointee.parm.0)

        // SET COLLISION STUFF

        newObj.pointee.CType = UInt32(CTYPE_SOLIDTOENEMY | CTYPE_PLAYERTEST | CTYPE_WEAPONTEST)
        newObj.pointee.CBits = UInt32(CBITS_ALLSOLID)
        CreateCollisionBoxFromBoundingBox_Rotated(newObj, 1, 1)

        return 1 // item was added
    }
}

// MARK: - Move gas mound

private let ventOff: [OGLPoint3D] = [
    OGLPoint3D(x: 2.163, y: 125.091, z: -1.907),
    OGLPoint3D(x: 19.385, y: 69.243, z: 2.243),
    OGLPoint3D(x: 5.189, y: 22.168, z: -5.504),
]

private let cMoveGasMound: @convention(c) (UnsafeMutablePointer<ObjNode>?) -> Void = { theNodeOpt in
    let theNode = theNodeOpt!
    let fps = gFramesPerSecondFrac

    if TrackTerrainItem(theNode) != 0 { // just check to see if it's gone
        DeleteObject(theNode)
        return
    }

    // SEE IF SMOKE HITS PLAYER

    for i in 0..<Int(gNumPlayers) {
        if GetPlayerIsDead(Int32(i)) == 0 { // ingore dead players
            let pi = GetPlayerInfoEntry(Int32(i))
            var dist = CalcQuickDistance(theNode.pointee.Coord.x, theNode.pointee.Coord.z, pi.pointee.coord.x, pi.pointee.coord.z)

            if dist < 150.0 {
                dist = pi.pointee.coord.y - theNode.pointee.Coord.y
                if dist < 700.0 {
                    if !pi.pointee.objNode!.pointee.Skeleton!.isAnim(.disoriented) { // play effect on 1st hit
                        PlayEffect3D(Int16(EFFECT_BODYHIT), &pi.pointee.coord)
                        PlayRumbleEffect(Int16(EFFECT_BODYHIT), Int32(i))
                    }

                    _ = PlayerLoseHealth(Int16(i), fps * 0.1, .deathDive, nil, 1)
                }
            }
        }
    }

    // MAKE SMOKE

    theNode.pointee.ParticleTimer -= fps // see if add smoke
    if theNode.pointee.ParticleTimer <= 0.0 {
        theNode.pointee.ParticleTimer += 0.05 // reset timer

        var particleGroup = theNode.pointee.ParticleGroup
        let magicNum = theNode.pointee.ParticleMagicNum

        if particleGroup == -1 || VerifyParticleGroupMagicNum(particleGroup, magicNum) == 0 {
            let newMagicNum = MyRandomLong() // generate a random magic num
            theNode.pointee.ParticleMagicNum = newMagicNum

            gNewParticleGroupDef.magicNum = newMagicNum
            gNewParticleGroupDef.particleType = .fallingSparks
            gNewParticleGroupDef.flags = UInt32(PARTICLE_FLAGS_DONTCHECKGROUND)
            gNewParticleGroupDef.gravity = 100
            gNewParticleGroupDef.magnetism = 0
            gNewParticleGroupDef.baseScale = 15.0
            gNewParticleGroupDef.decayRate = -2.5
            gNewParticleGroupDef.fadeRate = 0.25
            gNewParticleGroupDef.particleTextureNum = UInt8(PARTICLE_SObjType_GasCloud)
            gNewParticleGroupDef.srcBlend = GL_SRC_ALPHA
            gNewParticleGroupDef.dstBlend = GL_ONE_MINUS_SRC_ALPHA
            particleGroup = NewParticleGroup(&gNewParticleGroupDef)
            theNode.pointee.ParticleGroup = particleGroup
        }

        if particleGroup != -1 {
            var p = ventOff[Int(theNode.pointee.Kind)].transformed(by: theNode.pointee.BaseTransformMatrix)

            let x = p.x
            let y = p.y
            let z = p.z

            for _ in 0..<2 {
                var d = OGLVector3D()

                p.x = x + RandomFloat2() * 15.0
                p.y = y + RandomFloat() * 10.0
                p.z = z + RandomFloat2() * 15.0

                d.x = RandomFloat2() * 30.0
                d.y = 300.0 + RandomFloat() * 150.0
                d.z = RandomFloat2() * 30.0

                var newParticleDef = NewParticleDefType()
                newParticleDef.groupNum = particleGroup
                newParticleDef.scale = RandomFloat() + 1.0
                newParticleDef.rotZ = RandomFloat() * SwPI2
                newParticleDef.rotDZ = RandomFloat2() * 2.0
                newParticleDef.alpha = 0.6

                let stop: Bool = withUnsafeMutablePointer(to: &p) { pPtr in
                    withUnsafeMutablePointer(to: &d) { dPtr in
                        newParticleDef.where = pPtr
                        newParticleDef.delta = dPtr
                        return AddParticleToGroup(&newParticleDef) != 0
                    }
                }
                if stop {
                    theNode.pointee.ParticleGroup = -1
                    break
                }
            }
        }
    }
}

// MARK: -

// MARK: - Trigger callback: misc smackable object

// Returns TRUE if want to handle hit as a solid
func DoTrig_MiscSmackableObject(_ trigger: UnsafeMutablePointer<ObjNode>!, _ theNode: UnsafeMutablePointer<ObjNode>!) -> UInt8 {
    if PlayerSmackedIntoObject(theNode, trigger, .explode) != 0 {
        return 0
    }

    return 1
}

// MARK: -

// MARK: - Add asteroid

extension UnsafeMutablePointer where Pointee == TerrainItemEntryType {
    @discardableResult
    func addAsteroid(x: Float, z: Float) -> UInt8 {
        var def = NewObjectDefinitionType()
        def.group = UInt8(MODEL_GROUP_LEVELSPECIFIC)
        def.type = UInt8(Int32(LEVEL3_ObjType_Asteroid_Cracked) + Int32(pointee.parm.0))
        def.scale = 4.0 + RandomFloat2() * 1.0
        def.coord.x = x
        def.coord.z = z
        def.flags = gAutoFadeStatusBits
        def.slot = 197
        def.moveCall = nil
        def.rot = RandomFloat() * SwPI2

        def.coord.y = GetMinTerrainY(x, z, Int16(def.group), Int16(def.type), def.scale) - GetObjectGroupBBox(Int32(def.group), Int32(def.type)).min.y

        let newObj = MakeNewDisplayGroupObject(&def)!

        newObj.pointee.TerrainItemPtr = self // keep ptr to item list

        newObj.pointee.Kind = Int32(pointee.parm.0)

        // SET COLLISION STUFF

        newObj.pointee.CType = UInt32(CTYPE_SOLIDTOENEMY | CTYPE_PLAYERTEST | CTYPE_WEAPONTEST)
        newObj.pointee.CBits = UInt32(CBITS_ALLSOLID)
        CreateCollisionBoxFromBoundingBox_Rotated(newObj, 1, 1)

        return 1 // item was added
    }
}
